SSHAgent issue after upgrading to 5.1

**Describe the bug**

(i'm sorry i posted twice on rundeck-discuss but my message seems to not be sent)

i accidently upgraded to 5.1 from 5.0 (apt repo) and now no nodes are dispatched with this following error :

```
Failed: PluginFailed: com/dtolabs/rundeck/core/utils/SSHAgentProcess
Execution failed: 4999510 in project Qualif: [Workflow result: , step failures: {1=Dispatch failed on 1 nodes: [taskq.local: PluginFailed: com/dtolabs/rundeck/core/utils/SSHAgentProcess + {dataContext=MultiDataContextImpl(map={}, base=null)} ]}, Node failures: {taskq.local=[PluginFailed: com/dtolabs/rundeck/core/utils/SSHAgentProcess + 
{dataContext=MultiDataContextImpl(map={}, base=null)} ]}, status: failed]
```

**My Rundeck detail**

* Rundeck version: 5.1
* install type: apt repo
* OS Name/version: ubuntu 20.04
* DB Type/version: mysql

**To Reproduce**
Steps to reproduce the behavior:
1. upgraded from 5.0.0 to 5.1

**Expected behavior**
nodes are dispatched without issue

**Additional context**



rundeck is running on ubuntu 20.04 with mariadb
ansible 2.9.6
executor : ansible
file copier : ansible

thank you for you help !


here is the complete debug trace :


```
[workflow] beginExecuteNodeStep(taskq.local): NodeDispatch: StepExecutionItem{type='NodeDispatch', keepgoingOnSuccess=false, hasFailureHandler=false}
Failed executing node plugin [exec-command] on node taskq.local: java.lang.NoClassDefFoundError: com/dtolabs/rundeck/core/utils/SSHAgentProcess
	at com.rundeck.plugins.ansible.ansible.AnsibleRunner.registerKeySshAgent(AnsibleRunner.java:624)
	at com.rundeck.plugins.ansible.ansible.AnsibleRunner.run(AnsibleRunner.java:456)
	at com.rundeck.plugins.ansible.plugin.AnsibleNodeExecutor.executeCommand(AnsibleNodeExecutor.java:179)
	at com.dtolabs.rundeck.core.execution.ExecutionServiceImpl.executeCommand(ExecutionServiceImpl.java:427)
	at com.dtolabs.rundeck.core.execution.ExecutionService$executeCommand.call(Unknown Source)
	at org.rundeck.plugin.scriptnodestep.CommandNodeStepPlugin.executeNodeStep(CommandNodeStepPlugin.groovy:49)
	at com.dtolabs.rundeck.core.execution.workflow.steps.node.NodeStepPluginAdapter.executeNodeStep(NodeStepPluginAdapter.java:169)
	at com.dtolabs.rundeck.core.execution.ExecutionServiceImpl.executeNodeStep(ExecutionServiceImpl.java:207)
	at com.dtolabs.rundeck.core.execution.dispatch.SequentialNodeDispatcher.dispatch(SequentialNodeDispatcher.java:130)
	at com.dtolabs.rundeck.core.execution.dispatch.SequentialNodeDispatcher.dispatch(SequentialNodeDispatcher.java:61)
	at com.dtolabs.rundeck.core.execution.ExecutionServiceImpl.dispatchToNodesWith(ExecutionServiceImpl.java:263)
	at com.dtolabs.rundeck.core.execution.ExecutionServiceImpl.dispatchToNodes(ExecutionServiceImpl.java:234)
	at com.dtolabs.rundeck.core.execution.workflow.steps.NodeDispatchStepExecutor.executeWorkflowStep(NodeDispatchStepExecutor.java:66)
	at com.dtolabs.rundeck.core.execution.ExecutionServiceImpl.executeStep(ExecutionServiceImpl.java:111)
	at com.dtolabs.rundeck.core.execution.workflow.BaseWorkflowExecutor.executeWFItem(BaseWorkflowExecutor.java:285)
	at com.dtolabs.rundeck.core.execution.workflow.BaseWorkflowExecutor.executeWorkflowStep(BaseWorkflowExecutor.java:681)
	at com.dtolabs.rundeck.core.execution.workflow.engine.StepCallable.apply(StepCallable.java:71)
	at com.dtolabs.rundeck.core.execution.workflow.engine.StepOperation.apply(StepOperation.java:76)
	at com.dtolabs.rundeck.core.execution.workflow.engine.StepOperation.apply(StepOperation.java:32)
	at com.dtolabs.rundeck.core.rules.WorkflowEngineOperationsProcessor.lambda$beginOperation$1(WorkflowEngineOperationsProcessor.java:323)
	at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:131)
	at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:75)
	at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:82)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: java.lang.ClassNotFoundException: com.dtolabs.rundeck.core.utils.SSHAgentProcess
	... 26 more
Failed: PluginFailed: com/dtolabs/rundeck/core/utils/SSHAgentProcess
[workflow] finishExecuteNodeStep(taskq.local): NodeDispatch: PluginFailed: com/dtolabs/rundeck/core/utils/SSHAgentProcess
1: Workflow step finished, result: Dispatch failed on 1 nodes: [taskq.local: PluginFailed: com/dtolabs/rundeck/core/utils/SSHAgentProcess + {dataContext=MultiDataContextImpl(map={}, base=null)} ]
[workflow] Finish step: 1,NodeDispatch
[wf:5b4a53c7-892a-4892-8509-d35101bcf3a3] OperationFailed: operation completed, success? false: OperationCompleted(identity=[1], stepNum=1, newState=DataState{state={step.1.result.failedNodes=taskq.local, step.1.completed=true, step.any.state.failed=true, before.step.1=false, step.1.state=failure, after.step.1=true}}, stepResultCapture=StepResultCapture{stepResult=Dispatch failed on 1 nodes: [taskq.local: PluginFailed: com/dtolabs/rundeck/core/utils/SSHAgentProcess + {dataContext=MultiDataContextImpl(map={}, base=null)} ], stepSuccess=false, statusString='null', controlBehavior=null, resultData=MultiDataContextImpl(map={}, base=null)}, success=false)
[wf:5b4a53c7-892a-4892-8509-d35101bcf3a3] WillProcessStateChange: state changes: [1] DataState{state={step.1.result.failedNodes=taskq.local, step.1.completed=true, step.any.state.failed=true, before.step.1=false, step.1.state=failure, after.step.1=true}}
[wf:5b4a53c7-892a-4892-8509-d35101bcf3a3] Update conditional state: {step.1.result.failedNodes=taskq.local, step.1.completed=true, step.any.state.failed=true, before.step.1=false, step.1.state=failure, after.step.1=true}
[wf:5b4a53c7-892a-4892-8509-d35101bcf3a3] Update conditional state: {workflow.done=true, step.1.start=true, step.1.skip=true}
[wf:5b4a53c7-892a-4892-8509-d35101bcf3a3] DidProcessStateChange: applied state changes and rules (changed? true): [1] - StateLogger{state=DataState{state={workflow.id=5b4a53c7-892a-4892-8509-d35101bcf3a3, step.1.result.failedNodes=taskq.local, workflow.keepgoing=false, workflow.done=true, step.1.completed=true, step.any.state.failed=true, before.step.1=false, step.1.start=true, after.step.1=true, step.1.state=failure, workflow.state=started, step.1.skip=true}}}
[wf:5b4a53c7-892a-4892-8509-d35101bcf3a3] WorkflowEndState: Workflow end state reached.
[wf:5b4a53c7-892a-4892-8509-d35101bcf3a3] WillShutdown: Workflow engine shutting down (interrupted? false)
[wf:5b4a53c7-892a-4892-8509-d35101bcf3a3] Complete: Workflow complete: [Step{stepNum=1, label='null'}: OperationCompleted(identity=[1], stepNum=1, newState=DataState{state={step.1.result.failedNodes=taskq.local, step.1.completed=true, step.any.state.failed=true, before.step.1=false, step.1.state=failure, after.step.1=true}}, stepResultCapture=StepResultCapture{stepResult=Dispatch failed on 1 nodes: [taskq.local: PluginFailed: com/dtolabs/rundeck/core/utils/SSHAgentProcess + {dataContext=MultiDataContextImpl(map={}, base=null)} ], stepSuccess=false, statusString='null', controlBehavior=null, resultData=MultiDataContextImpl(map={}, base=null)}, success=false)]
[workflow] Finish execution:  node-first: [Workflow result: , step failures: {1=Dispatch failed on 1 nodes: [taskq.local: PluginFailed: com/dtolabs/rundeck/core/utils/SSHAgentProcess + {dataContext=MultiDataContextImpl(map={}, base=null)} ]}, Node failures: {taskq.local=[PluginFailed: com/dtolabs/rundeck/core/utils/SSHAgentProcess + {dataContext=MultiDataContextImpl(map={}, base=null)} ]}, status: failed]
[Workflow result: , step failures: {1=Dispatch failed on 1 nodes: [taskq.local: PluginFailed: com/dtolabs/rundeck/core/utils/SSHAgentProcess + {dataContext=MultiDataContextImpl(map={}, base=null)} ]}, Node failures: {taskq.local=[PluginFailed: com/dtolabs/rundeck/core/utils/SSHAgentProcess + {dataContext=MultiDataContextImpl(map={}, base=null)} ]}, status: failed]
Execution failed: 4999511 in project Moodle-Qualif: [Workflow result: , step failures: {1=Dispatch failed on 1 nodes: [taskq.local: PluginFailed: com/dtolabs/rundeck/core/utils/SSHAgentProcess + {dataContext=MultiDataContextImpl(map={}, base=null)} ]}, Node failures: {taskq.local=[PluginFailed: com/dtolabs/rundeck/core/utils/SSHAgentProcess + {dataContext=MultiDataContextImpl(map={}, base=null)} ]}, status: failed]
```

Reproduced.

It works on 4.17.X /  5.0.X but on 5.1 doesn't.

Steps to reproduce it:

1. Create an SSH key pair with a passphrase to connect to the remote nodes.
2. Create an Ansible Project, and Configure the Ansible Model Source (define an authentication method, username, SSH private key path, and the SSH passphrase on the Model Source SSH Connection Section).
3. Click on the nodes page.
4. Check the service.log (also the nodes don't appear on the Nodes page).

On 5.1.0:

```
ava.lang.NoClassDefFoundError: com/dtolabs/rundeck/core/utils/SSHAgentProcess
	at com.rundeck.plugins.ansible.ansible.AnsibleRunner.registerKeySshAgent(AnsibleRunner.java:624) ~[?:?]
	at com.rundeck.plugins.ansible.ansible.AnsibleRunner.run(AnsibleRunner.java:456) ~[?:?]
	at com.rundeck.plugins.ansible.plugin.AnsibleResourceModelSource.getNodes(AnsibleResourceModelSource.java:380) ~[?:?]
	at com.dtolabs.rundeck.core.resources.ExceptionCatchingResourceModelSource.getNodes(ExceptionCatchingResourceModelSource.java:58) [rundeck-core-5.1.0-20240214.jar!/:?]
	at com.dtolabs.rundeck.core.resources.DelegateResourceModelSource.getNodes(DelegateResourceModelSource.java:35) [rundeck-core-5.1.0-20240214.jar!/:?]
	at com.dtolabs.rundeck.core.common.ProjectNodeSupport.getNodeSet(ProjectNodeSupport.java:137) [rundeck-core-5.1.0-20240214.jar!/:?]
	at com.dtolabs.rundeck.core.common.ProjectNodeSupport$ProjectNodesSource.getNodes(ProjectNodeSupport.java:358) [rundeck-core-5.1.0-20240214.jar!/:?]
	at com.dtolabs.rundeck.core.resources.ExceptionCatchingResourceModelSource.getNodes(ExceptionCatchingResourceModelSource.java:58) [rundeck-core-5.1.0-20240214.jar!/:?]
	at com.dtolabs.rundeck.core.resources.ResourceModelSource$getNodes.call(Unknown Source) [rundeck-core-5.1.0-20240214.jar!/:?]
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47) [groovy-3.0.19.jar!/:3.0.19]
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125) [groovy-3.0.19.jar!/:3.0.19]
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:130) [groovy-3.0.19.jar!/:3.0.19]
	at rundeck.services.nodes.CachedProjectNodes.reloadNodeSet(CachedProjectNodes.groovy:44) [classes!/:?]
	at rundeck.services.nodes.CachedProjectNodes$reloadNodeSet.call(Unknown Source) [classes!/:?]
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47) [groovy-3.0.19.jar!/:3.0.19]
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125) [groovy-3.0.19.jar!/:3.0.19]
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:130) [groovy-3.0.19.jar!/:3.0.19]
	at rundeck.services.NodeService$_loadNodes_closure4.doCall(NodeService.groovy:295) [classes!/:?]
	at rundeck.services.NodeService$_loadNodes_closure4.doCall(NodeService.groovy) [classes!/:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:107) [groovy-3.0.19.jar!/:3.0.19]
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323) [groovy-3.0.19.jar!/:3.0.19]
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:274) [groovy-3.0.19.jar!/:3.0.19]
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1030) [groovy-3.0.19.jar!/:3.0.19]
	at groovy.lang.Closure.call(Closure.java:427) [groovy-3.0.19.jar!/:3.0.19]
	at groovy.lang.Closure.call(Closure.java:406) [groovy-3.0.19.jar!/:3.0.19]
	at com.codahale.metrics.Timer.time(Timer.java:118) [metrics-core-4.2.18.jar!/:4.2.18]
	at org.grails.plugins.metricsweb.MetricService.withTimer(MetricService.groovy:61) [grails-metricsweb-5.1.0-20240214-plain.jar!/:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
	at org.codehaus.groovy.runtime.callsite.PlainObjectMetaMethodSite.doInvoke(PlainObjectMetaMethodSite.java:48) [groovy-3.0.19.jar!/:3.0.19]
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite$PogoCachedMethodSiteNoUnwrap.invoke(PogoMetaMethodSite.java:178) [groovy-3.0.19.jar!/:3.0.19]
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite.call(PogoMetaMethodSite.java:69) [groovy-3.0.19.jar!/:3.0.19]
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47) [groovy-3.0.19.jar!/:3.0.19]
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125) [groovy-3.0.19.jar!/:3.0.19]
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:157) [groovy-3.0.19.jar!/:3.0.19]
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callSafe(AbstractCallSite.java:112) [groovy-3.0.19.jar!/:3.0.19]
	at rundeck.services.NodeService$_loadNodes_closure5.doCall(NodeService.groovy:302) [classes!/:?]
	at rundeck.services.NodeService$_loadNodes_closure5.doCall(NodeService.groovy) [classes!/:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:107) [groovy-3.0.19.jar!/:3.0.19]
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323) [groovy-3.0.19.jar!/:3.0.19]
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:274) [groovy-3.0.19.jar!/:3.0.19]
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1030) [groovy-3.0.19.jar!/:3.0.19]
	at groovy.lang.Closure.call(Closure.java:427) [groovy-3.0.19.jar!/:3.0.19]
	at groovy.lang.Closure.call(Closure.java:406) [groovy-3.0.19.jar!/:3.0.19]
	at groovy.lang.Closure.run(Closure.java:498) [groovy-3.0.19.jar!/:3.0.19]
	at org.springframework.core.task.SimpleAsyncTaskExecutor$ConcurrencyThrottlingRunnable.run(SimpleAsyncTaskExecutor.java:286) [spring-core-5.3.27.jar!/:5.3.27]
	at java.lang.Thread.run(Thread.java:829) [?:?]
Caused by: java.lang.ClassNotFoundException: com.dtolabs.rundeck.core.utils.SSHAgentProcess
	at java.net.URLClassLoader.findClass(URLClassLoader.java:476) ~[?:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:594) ~[?:?]
	at org.springframework.boot.loader.LaunchedURLClassLoader.loadClass(LaunchedURLClassLoader.java:151) ~[rundeck-5.1.0-20240214.war:?]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:527) ~[?:?]
	at com.dtolabs.rundeck.core.plugins.LocalFirstClassLoader.loadClass(LocalFirstClassLoader.java:58) ~[rundeck-core-5.1.0-20240214.jar!/:?]
	at com.dtolabs.rundeck.core.plugins.LocalFirstClassLoader.loadClass(LocalFirstClassLoader.java:44) ~[rundeck-core-5.1.0-20240214.jar!/:?]
```

On 4.17.X / 5.0.X:

![Screenshot_2024-02-19_15-36-32](https://github.com/rundeck/rundeck/assets/51376003/a591153b-2619-43a8-a7ef-13b8a885b439)

Thanks a lot, @rimkov and @Elpatii.

--- a/core/src/main/java/com/dtolabs/rundeck/core/utils/SSHAgentProcess.java
+++ b/core/src/main/java/com/dtolabs/rundeck/core/utils/SSHAgentProcess.java
@@ -14,7 +14,7 @@
  * limitations under the License.
  */
 
-package org.rundeck.plugins.jsch.util;
+package com.dtolabs.rundeck.core.utils;
 
 import java.io.IOException;
 import java.io.InputStream;

--- a/plugins/jsch-plugin/src/main/java/org/rundeck/plugins/jsch/net/ExtSSHExec.java
+++ b/plugins/jsch-plugin/src/main/java/org/rundeck/plugins/jsch/net/ExtSSHExec.java
@@ -25,7 +25,7 @@
 
 package org.rundeck.plugins.jsch.net;
 
-import org.rundeck.plugins.jsch.util.SSHAgentProcess;
+import com.dtolabs.rundeck.core.utils.SSHAgentProcess;
 import com.dtolabs.rundeck.plugins.PluginLogger;
 import com.jcraft.jsch.*;
 import org.apache.tools.ant.BuildException;

--- a/plugins/jsch-plugin/src/main/java/org/rundeck/plugins/jsch/net/ExtScp.java
+++ b/plugins/jsch-plugin/src/main/java/org/rundeck/plugins/jsch/net/ExtScp.java
@@ -16,7 +16,7 @@
 
 package org.rundeck.plugins.jsch.net;
 
-import org.rundeck.plugins.jsch.util.SSHAgentProcess;
+import com.dtolabs.rundeck.core.utils.SSHAgentProcess;
 import com.dtolabs.rundeck.plugins.PluginLogger;
 import com.jcraft.jsch.JSchException;
 import com.jcraft.jsch.Session;

--- a/plugins/jsch-plugin/src/main/java/org/rundeck/plugins/jsch/net/SSHTaskBuilder.java
+++ b/plugins/jsch-plugin/src/main/java/org/rundeck/plugins/jsch/net/SSHTaskBuilder.java
@@ -26,7 +26,7 @@ package org.rundeck.plugins.jsch.net;
 import com.dtolabs.rundeck.core.common.INodeEntry;
 import com.dtolabs.rundeck.core.dispatcher.DataContextUtils;
 import com.dtolabs.rundeck.core.utils.FileUtils;
-import org.rundeck.plugins.jsch.util.SSHAgentProcess;
+import com.dtolabs.rundeck.core.utils.SSHAgentProcess;
 import com.dtolabs.rundeck.plugins.PluginLogger;
 import com.dtolabs.utils.Streams;
 import com.jcraft.jsch.JSch;

--- a/plugins/jsch-plugin/src/test/java/org/rundeck/plugins/jsch/net/TestSSHTaskBuilder.java
+++ b/plugins/jsch-plugin/src/test/java/org/rundeck/plugins/jsch/net/TestSSHTaskBuilder.java
@@ -25,7 +25,7 @@ package org.rundeck.plugins.jsch.net;
 
 import com.dtolabs.rundeck.core.cli.CLIUtils;
 import com.dtolabs.rundeck.core.common.NodeEntryImpl;
-import org.rundeck.plugins.jsch.util.SSHAgentProcess;
+import com.dtolabs.rundeck.core.utils.SSHAgentProcess;
 import com.dtolabs.rundeck.plugins.PluginLogger;
 
 import junit.framework.TestCase;

